client/verta/verta/_internal_utils/_utils.py

```diff
@@ -68,19 +72,89 @@ def __init__(self, scheme=None, socket=None, auth=None, max_retries=0, ignore_co
             on HTTP codes {502, 503, 504} which commonly occur during back end connection lapses.
         ignore_conn_err : bool, default False
             Whether to ignore connection errors and instead return successes with empty contents.
-
+        credentials : :class:`~verta._internal_utils.credentials.EmailCredentials` or :class:`~verta._internal_utils.credentials.JWTCredentials`, optional
+            Either dev key or JWT token data to be used for authentication
+        headers: dict, optional
+            Additional headers to attach to requests
         """
+        self._init_headers()
         self.scheme = scheme
         self.socket = socket
-        self.auth = auth
         # TODO: retry on 404s, but only if we're sure it's not legitimate e.g. from a GET
         self.retry = Retry(total=max_retries,
                            backoff_factor=1,  # each retry waits (2**retry_num) seconds
                            method_whitelist=False,  # retry on all HTTP methods
-                           status_forcelist=(502, 503, 504),  # only retry on these status codes
+                           status_forcelist=(http_codes.bad_gateway, http_codes.unavailable, http_codes.gateway_timeout),  # only retry on these status codes
                            raise_on_redirect=False,  # return Response instead of raising after max retries
                            raise_on_status=False)  # return Response instead of raising after max retries
         self.ignore_conn_err = ignore_conn_err
+        self.credentials = credentials
+        self.headers = headers
+
+
+    @property
+    def credentials(self):
+        return self._credentials
+
+    @credentials.setter
+    def credentials(self, value):
+        self._credentials = value
+        self._auth_headers = self._headers_for_credentials(value)
```

Comment:
I'd personally prefer having `_auth_headers` be a property that calculates the headers from `self._credentials` when accessed than tracking it as separate, duplicated-ish state.

Reply:
Good point. Removed `_auth_headers`.